framework/src/kernels/ADKernel.C

```diff
@@ -167,12 +167,24 @@ template <typename T, ComputeStage compute_stage>
 void
 ADKernelTempl<T, compute_stage>::computeADOffDiagJacobian()
 {
-  std::vector<DualReal> residuals(_test.size(), 0);
+  if (_residuals.size() != _test.size())
+    _residuals.resize(_test.size(), 0);
+  for (auto & r : _residuals)
+    r = 0;
 
   precalculateResidual();
+  ADResidual tmp;
   for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    tmp = _ad_JxW[_qp];
+    tmp *= _ad_coord[_qp];
     for (_i = 0; _i < _test.size(); _i++)
-      residuals[_i] += _ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual();
+    {
+      _r = tmp;
+      _r *= computeQpResidual();
+      _residuals[_i] += _r;
```

Comment:
this is better than

```
_residuals[_i] += tmp * computeQpResidual();
```

??? Are compilers that dumb? It sure is more painful to look at.

Reply:
I’m with @dschwen .  His suggestion is actually a multiply-add which should really be better.

BUT: maybe @rwcarlsen went with this to remove temporaries... which are very costly.

Reply:
Yeah - this wasn't a clear win.  I just took another look at some profiles and I think this is not worth it. Fixed.